### ISSUE TITLE
CA1823 erroneously fires when accesses are all from static methods

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#2776]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.4 (Latest)

#### Diagnostic ID

Example: [CA1823](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1823-avoid-unused-private-fields)

#### Repro steps
```C#
#nullable enable
class Program
{
    static void Main() => SomeClass<int>.Return(SomeClass<int>.GetOrCreate());
}


public class SomeClass<TInner>
{
    private static SomeClass<TInner>? s_cache;
    private SomeClass<TInner>? _next;

    internal static SomeClass<TInner> GetOrCreate()
    {
        SomeClass<TInner>? c = s_cache;
        if (c != null)
        {
            s_cache = c._next;
            c._next = null;
            return c;
        }

        return new SomeClass<TInner>();
    }

    internal static void Return(SomeClass<TInner> c)
    {
        c._next = s_cache;
        s_cache = c;
    }
}
```

#### Expected behavior
No CA1823 warnings.

#### Actual behavior
CA1823 fires saying:
```
warning CA1823: Unused field '_next'
```
This is obviously incorrect.

**Comment by mavasani** [comment]
@stephentoub Does the analyzer fire only when nullable is enabled? Tagging @jasonmalinowski @ryzngard

**Comment by stephentoub** [comment]
> Does the analyzer fire only when nullable is enabled?

Yes.  The error goes away if I remove the `#nullable enable` and the three `?`s.  It also goes away if I add any reference to `_next` that doesn't specify the target instance.

**Comment by mavasani** [comment]
Also tagging @333fred in case he can think of anything obvious that might be causing this. A quick glance at the [analyzer code](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidUnusedPrivateFields.cs) does not show any type equality checks being done, which would have been my first guess for this behavior change...

**Comment by mavasani** [comment]
Closer look at code indicates that the analyzer might be relying on some ordering behavior between FieldSymbol callback for the field declaration and FieldReference callback. The TryAdd and TryRemove operations on `maybeUnreferencedPrivateFields` field in the analyzer seem to have a race condition associated with it. Likely the analyzer callback ordering got changed in presence of nullability. I will fix the analyzer.

**Comment by mavasani** [comment]
Okay, all my analysis from code inspection was incorrect. There is no race condition in the analyzer, but it seems to be a breaking change for analyzers in presence of analyzers. The `IFieldSymbol` being passed to the symbol action callback for field declaration is different from the symbol exposed from `IFieldReferenceOperation.Field` - which leads to analyzer not finding any references to declared fields. @333fred do we know about this? Seems like a bad breaking change...

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn/issues/38195, which is the root cause of analyzer behavior change. @333fred any workaround that you feel can me made from analyzer side would also be fine (like using a different symbol equality comparer).
